libtenzir/builtins/formats/feather2.cpp

```diff
@@ -0,0 +1,203 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <tenzir/argument_parser.hpp>
+#include <tenzir/make_byte_reader.hpp>
+#include <tenzir/plugin.hpp>
+#include <tenzir/table_slice.hpp>
+
+#include <arrow/array.h>
+#include <arrow/compute/cast.h>
+#include <arrow/io/api.h>
+#include <arrow/ipc/api.h>
+#include <arrow/ipc/reader.h>
+#include <arrow/ipc/writer.h>
+#include <arrow/record_batch.h>
+#include <arrow/table.h>
+#include <arrow/util/key_value_metadata.h>
+#include <caf/expected.hpp>
+
+#include <queue>
+
+namespace tenzir::plugins::feather2 {
+namespace {
+
+class CallbackListener : public arrow::ipc::Listener {
+public:
+  CallbackListener() = default;
+
+  arrow::Status OnRecordBatchDecoded(
+    std::shared_ptr<arrow::RecordBatch> record_batch) override {
+    record_batch_buffer.push(record_batch);
+    return arrow::Status::OK();
+  }
+
+  std::queue<std::shared_ptr<arrow::RecordBatch>> record_batch_buffer;
+};
+
+auto parse_feather(generator<chunk_ptr> input, operator_control_plane& ctrl)
+  -> generator<table_slice> {
+  auto byte_reader = make_byte_reader(std::move(input));
+  auto listener = std::make_shared<CallbackListener>();
+  auto stream_decoder = arrow::ipc::StreamDecoder(listener);
+
+  while (true) {
+    auto required_size
+      = detail::narrow_cast<size_t>(stream_decoder.next_required_size());
+    auto payload = byte_reader(required_size);
+
+    if (!payload) {
+      co_yield {};
+      continue;
+    }
+
+    const auto done = payload->size() < required_size;
+    auto decode_result
+      = stream_decoder.Consume(as_arrow_buffer(std::move(payload)));
+    if (!decode_result.ok()) {
+      diagnostic::error("failed to decode the byte stream into a record batch")
+        .note("{}", decode_result.ToString())
+        .emit(ctrl.diagnostics());
+      co_return;
+    }
+
+    while (!listener->record_batch_buffer.empty()) {
+      auto batch = listener->record_batch_buffer.front();
+      listener->record_batch_buffer.pop();
+      auto validate_status = batch->Validate();
+      TENZIR_ASSERT(validate_status.ok(), validate_status.ToString().c_str());
+      co_yield table_slice(batch);
```

Comment:
What if this fails?

The `table_slice` constructor from a record batch currently assumes that all data is compatible with Tenzir's data format. Can we change that to not crash but rather error if there's a mismatch?

Reply:
We created a follow-up issue: https://github.com/tenzir/issues/issues/1715